Ical.Net/Evaluation/RecurringEvaluator.cs

```diff
@@ -112,49 +107,50 @@ protected HashSet<Period> EvaluateExRule(IDateTime referenceDate, DateTime perio
     /// <param name="periodEnd">The end date of the range to evaluate.</param>
-    protected HashSet<Period> EvaluateExDate(IDateTime referenceDate, DateTime periodStart, DateTime periodEnd)
+    protected IEnumerable<Period> EvaluateExDate(IDateTime referenceDate, DateTime? periodStart, DateTime? periodEnd)
     {
         if (Recurrable.ExceptionDates == null || !Recurrable.ExceptionDates.Any())
-        {
-            return new HashSet<Period>();
-        }
+            return [];
 
-        var exDates = new HashSet<Period>(Recurrable.ExceptionDates.SelectMany(exDate => exDate));
+        var exDates = new SortedSet<Period>(Recurrable.ExceptionDates.SelectMany(exDate => exDate));
         return exDates;
     }
 
-    public override HashSet<Period> Evaluate(IDateTime referenceDate, DateTime periodStart, DateTime periodEnd, bool includeReferenceDateInResults)
+    public override IEnumerable<Period> Evaluate(IDateTime referenceDate, DateTime? periodStart, DateTime? periodEnd, bool includeReferenceDateInResults)
     {
-        var periods = new HashSet<Period>();
-
         var rruleOccurrences = EvaluateRRule(referenceDate, periodStart, periodEnd, includeReferenceDateInResults);
         //Only add referenceDate if there are no RecurrenceRules defined
         if (includeReferenceDateInResults && (Recurrable.RecurrenceRules == null || !Recurrable.RecurrenceRules.Any()))
         {
-            rruleOccurrences.UnionWith(new[] { new Period(referenceDate), });
+            rruleOccurrences = rruleOccurrences.Append(new Period(referenceDate));
         }
 
         var rdateOccurrences = EvaluateRDate(referenceDate, periodStart, periodEnd);
 
         var exRuleExclusions = EvaluateExRule(referenceDate, periodStart, periodEnd);
         var exDateExclusions = EvaluateExDate(referenceDate, periodStart, periodEnd);
 
-        //Exclusions trump inclusions
-        periods.UnionWith(rruleOccurrences);
-        periods.UnionWith(rdateOccurrences);
-        periods.ExceptWith(exRuleExclusions);
-        periods.ExceptWith(exDateExclusions);
-
-        var dateOverlaps = FindDateOverlaps(periods, exDateExclusions);
-        periods.ExceptWith(dateOverlaps);
+        var periods =
+            rruleOccurrences
+            .OrderedMerge(rdateOccurrences)
+            .OrderedDistinct()
+            .OrderedExclude(exRuleExclusions)
+            .OrderedExclude(exDateExclusions, Comparer<Period>.Create(CompareDateOverlap));
 
         return periods;
     }
 
-    private static HashSet<Period> FindDateOverlaps(HashSet<Period> periods, HashSet<Period> dates)
+    /// <summary>
+    /// Compares whether the given period's date overlaps with the given EXDATE. The dates are
+    /// considered to overlap if they start at the same time, or the EXDATE is an all-day date
+    /// and the period's start date is the same as the EXDATE's date.
+    /// </summary>
+    private static int CompareDateOverlap(Period period, Period exDate)
```

Comment:
exDate still a period unfortunately, but that's for another PR

Reply:
Yes. As the PR is quite big, I tried to leave things the way they are, where possible (or implemented PRs upfront like #659), but that's not always obvious from the diff.